### ISSUE TITLE
Attributes from Samplers not being added to sampled Spans

### ISSUE THREAD

**Issue by paulosman** [open-telemetry/opentelemetry-go#1228]
The [sampling section of the Tracing SDK specification](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#sampling) specifies that the return value of the ShouldSample method contains "A set of span Attributes that will also be added to the Span.". 

When using a custom sampler with the Go SDK, attributes returned from the Sampler are only set on the root Span in a Trace, not on each of the individual spans. I tested with the Python SDK, and in that implementation, Attributes are being set on every span. 

The use case we need this for is to be able to set the Sample Rate for each span sent to Honeycomb. Our backend uses the sample rate when calculating aggregate functions on events. Here's the [sampler implementation we're using](https://github.com/honeycombio/opentelemetry-samplers-go/blob/main/honeycombsamplers/deterministic_sampler.go#L58) for a proof of concept. 

Using [this sample app](https://github.com/honeycombio/otel-sampling-demos/blob/master/golang/deterministic/main.go), we get the following output:

```
[
  {
    "SpanContext": {
    ...
    },
    "Attributes": null,
  }
]
```

On each of the Spans, except to the root span:

```
[
  {
     "SpanContext": {
      ...
  },
  "Attributes": [
    {
      "Key": "SampleRate",
      "Value": {
        "Type": "INT32",
        "Value": 2
      }
    },
    ...
    ]
  }
]
```

Using the [Python sample app](https://github.com/honeycombio/otel-sampling-demos/blob/master/python/deterministic/app.py), I'm able to see the Attribute on every span:

```
    ...
    "attributes": {
        "SampleRate": 2
    },
    ...
```

**Comment by paulosman** [comment]
btw, I'm happy to take a crack at a PR for this but want to make sure others agree with the framing of the problem first.

**Comment by Aneurysm9** [comment]
I can't access [the sample application](https://github.com/honeycombio/otel-sampling-demos/blob/master/golang/deterministic/main.go), perhaps the repo is private?

It looks like we might be [slightly overzealous](https://github.com/open-telemetry/opentelemetry-go/blob/master/sdk/trace/span.go#L401) in attempting to invoke samplers.  This may be a holdover from before [the `ParentBased` sampler was made significantly more flexible](https://github.com/open-telemetry/opentelemetry-go/pull/1153).  Can you confirm that the issue is in fact that your sampler is not being invoked at all for child spans rather than the attributes not being set?

**Comment by paulosman** [comment]
My apologies, the [repo is public](https://github.com/honeycombio/otel-sampling-demos/blob/master/golang/deterministic/main.go) now in case it's still helpful. 

I've confirmed - the `ShouldSample` method is only being called once per trace, and not once per span. I tried commenting out that guard that you linked to and indeed that results in the expected behavior - the sampler is consulted for each span and attributes are set properly.
